### PR TITLE
check plugin manifest is correctly formatted before push

### PR DESCRIPTION
**Pull request by vieux**
try to `json.Unmarshal` the manifest before pushing it the the registry to validate its format.

maybe there is a better way to do that ? /cc @crosbymichael ?

ping @tiborvass @anusha-ragunathan

### CODE REVIEW

File: plugin/distribution/push.go [hunk 1]
```diff
@@ -72,13 +74,36 @@ func Push(name string, rs registry.Service, metaHeader http.Header, authConfig *
 		}
 		h := sha256.New()
 		r := io.TeeReader(f, h)
-		_, err = io.Copy(bw, r)
-		if err != nil {
+		if i == 0 {
+			b := &bytes.Buffer{}
+			_, err = io.Copy(b, r)
```
Comment (tiborvass) [inline plugin/distribution/push.go]: why do you need to copy in a buffer? Can't you pass `r` directly to Unmarshal?

**Comment by crosbymichael** [comment]
Are you just wanting to check that it is json or has the manifest format?

**Comment by vieux** [comment]
@crosbymichael just json I think

**Comment by crosbymichael** [comment]
I don't know of a better way to just check that some stream is json other than trying to unmarshal it into an interface/map interface

**Comment by vieux** [comment]
@tiborvass PTAL

**Comment by vdemeester** [comment]
LGTM :cat:

File: plugin/backend.go [hunk 2]
```diff
@@ -120,6 +121,16 @@ func (pm *Manager) Push(name string, metaHeader http.Header, authConfig *types.A
 	}
 	defer config.Close()
 
+	var dummy map[string]interface{}
+	err = json.NewDecoder(config).Decode(&dummy)
```
Comment (tonistiigi) [inline plugin/backend.go]: This doesn't catch garbage after the first JSON object. Like `{}{}`. As we don't actually use the reader anymore, it is probably easier to just read the whole file, unmarshal and then wrap it with `bytes.Buffer` before calling `Push`.

File: plugin/backend.go [hunk 3]
```diff
@@ -120,6 +121,16 @@ func (pm *Manager) Push(name string, metaHeader http.Header, authConfig *types.A
 	}
 	defer config.Close()
 
+	var dummy map[string]interface{}
```
Comment (tonistiigi) [inline plugin/backend.go]: Why not unmarshal to `types.Plugin`? We could even validate the type for some required keys.

**Comment by vieux** [comment]
@tonistiigi PTAL

File: plugin/backend.go [hunk 4]
```diff
@@ -114,19 +118,24 @@ func (pm *Manager) Push(name string, metaHeader http.Header, authConfig *types.A
...
-	defer config.Close()
 
 	rootfs, err := archive.Tar(filepath.Join(dest, "rootfs"), archive.Gzip)
 	if err != nil {
 		return err
 	}
 	defer rootfs.Close()
 
-	_, err = distribution.Push(name, pm.registryService, metaHeader, authConfig, config, rootfs)
+	_, err = distribution.Push(name, pm.registryService, metaHeader, authConfig, ioutils.NopReadCloser(bytes.NewReader(config)), rootfs)
```
Comment (tonistiigi) [inline plugin/backend.go]: `ioutil.NopCloser`

Comment (vieux) [inline plugin/backend.go]: 👼

**Comment by tonistiigi** [comment]
LGTM

**Comment by tiborvass** [comment]
LGTM, TIL something thanks @tonistiigi :)

**Pull request merged**